### ISSUE TITLE
E0277 false positive in batch diagnostics: `MutexGuard: Send` reported for spawned async fn with no guard held across await

### ISSUE THREAD

**Issue by fglogan** [rust-lang/rust-analyzer#23102]
`rust-analyzer diagnostics .` reports an E0277 hard error on code that rustc accepts. No `MutexGuard` is ever held across an await point.

### Minimal reproduction

`Cargo.toml`:

```toml
[package]
name = "repro-mutexguard-send"
version = "0.1.0"
edition = "2021"

[dependencies]
tokio = { version = "1", features = ["rt", "sync", "time", "macros"] }
```

`src/lib.rs`:

```rust
use std::path::PathBuf;
use std::sync::{Arc, Mutex};

pub struct WorkspaceState {
    pub value: u64,
}

pub async fn actor_body(
    path: PathBuf,
    workspace: Arc<Mutex<WorkspaceState>>,
    notify: Arc<tokio::sync::Notify>,
    last_error: Arc<Mutex<String>>,
) {
    let mut previous: Option<Vec<u8>> = None;
    loop {
        notify.notified().await;
        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
        let _ = &path;
        {
            let guard = workspace.lock().unwrap();
            let bytes = guard.value.to_le_bytes().to_vec();
            if previous.as_deref() != Some(bytes.as_slice()) {
                previous = Some(bytes);
            }
        }
        {
            let mut err = last_error.lock().unwrap();
            err.clear();
        }
        let _ = &previous;
    }
}

pub fn spawn_supervisor(
    path: PathBuf,
    workspace: Arc<Mutex<WorkspaceState>>,
    notify: Arc<tokio::sync::Notify>,
    last_error: Arc<Mutex<String>>,
) {
    tokio::spawn(async move {
        loop {
            let inner_path = path.clone();
            let inner_workspace = Arc::clone(&workspace);
            let inner_notify = Arc::clone(&notify);
            let inner_last = Arc::clone(&last_error);
            let handle = tokio::spawn(actor_body(
                inner_path,
                inner_workspace,
                inner_notify,
                inner_last,
            ));
            let _ = handle.await;
        }
    });
}
```

### Actual

```
Error RustcHardError("E0277") from LineCol { line: 52, col: 25 } to LineCol { line: 52, col: 37 }:
the trait bound `MutexGuard<'<erased>, WorkspaceState>: Send` is not satisfied
```

The span is the `Arc::clone(&last_error)` expression inside the supervisor loop (which contains no `lock()` call at all). Every lock scope in `actor_body` ends before the next await point, so the spawned futures are `Send`; `cargo check` and `cargo clippy --all-targets -- -D warnings` both accept the crate.

### Expected

No diagnostic.

### Versions

Reproduced with both:

- `rust-analyzer 1.97.1 (8bab26f4 2026-07-14)` (rustup component, stable toolchain)
- `rust-analyzer 0.3.2997-standalone (b54a82b321 2026-08-02)` (2026-08-03 release)

macOS aarch64. rustc 1.97.1.

### Possibly related symptom

In the larger workspace this was reduced from, the same two analyzer versions also report `the trait bound `Arguments<'<erased>>: Sync` is not satisfied` spanning two `axum` `WebSocketUpgrade::on_upgrade(move |socket| ...)` call chains whose handlers rustc likewise accepts. I could not reduce that one to a standalone crate (it seems to need the cross-crate async call graph), but the `'<erased>` lifetime in an auto-trait bound and the async-capture context match this report, so they may share a root cause in how auto traits are solved for async generator captures.

**Comment by ChayimFriedman2** [comment]
That's weird because we don't lower async, I'd expect every future to be `Send` without considering what it keeps across await points.
